listenbrainz/db/playlist.py

```diff
@@ -359,6 +366,174 @@ def get_playlists_collaborated_on(db_conn, ts_conn, user_id: int, include_privat
 
 
+def search_playlists_for_user(db_conn, ts_conn, user_id: int, query: str, count: int = 0, offset: int = 0):
+    """
+    Search for playlists by name or description
+
+    Arguments:
+        db_conn: database connection
+        ts_conn: timescale database connection
+        user_id: The user id
+        query: The search query
+        count: Return this many playlists. If 0, return all playlists
+        offset: if set, get playlists from this offset
+
+    Returns:
+        a tuple (playlists, total_playlists)
+    """
+    if count == 0:
+        count = None
+
+    params = {"query": query, "count": count, "offset": offset, "user_id": user_id}
+    query = text(f"""
+    WITH playlist_similarities AS (
+        SELECT pl.id
+             , pl.mbid
+             , pl.creator_id
+             , pl.name
+             , pl.description
+             , pl.public
+             , pl.created
+             , pl.last_updated
+             , pl.copied_from_id
+             , pl.created_for_id
+             , pl.additional_metadata
+             , copy.mbid as copied_from_mbid
+             , similarity(pl.name, :query) AS name_similarity
+             , similarity(pl.description, :query) AS description_similarity
+          FROM playlist.playlist AS pl
+     LEFT JOIN playlist.playlist AS copy
+            ON pl.copied_from_id = copy.id
+     LEFT JOIN playlist.playlist_collaborator
+            ON pl.id = playlist_collaborator.playlist_id
+        WHERE pl.creator_id = :user_id
+            OR pl.created_for_id = :user_id
+            OR playlist.playlist_collaborator.collaborator_id = :user_id
+            OR pl.public = true
+    )
+    SELECT *
+      FROM playlist_similarities
+     WHERE name_similarity > 0.1
+        OR description_similarity > 0.1
+  ORDER BY name_similarity DESC, description_similarity DESC
+     LIMIT :count
+    OFFSET :offset
+    """)
+
+    result = ts_conn.execute(query, params)
+    playlists = _playlist_resultset_to_model(db_conn, ts_conn, result, False)
+
+    # Fetch the total count of playlists
+    total_count = 0
+    query = text(f"""
+    WITH playlist_similarities AS (
+        SELECT similarity(pl.name, :query) AS name_similarity
+             , similarity(pl.description, :query) AS description_similarity
+          FROM playlist.playlist AS pl
+     LEFT JOIN playlist.playlist AS copy
+            ON pl.copied_from_id = copy.id
+     LEFT JOIN playlist.playlist_collaborator
+            ON pl.id = playlist_collaborator.playlist_id
+        WHERE pl.creator_id = :user_id
+            OR pl.created_for_id = :user_id
+            OR playlist.playlist_collaborator.collaborator_id = :user_id
+            OR pl.public = true
+    )
+    SELECT COUNT(*)
+      FROM playlist_similarities
+     WHERE name_similarity > 0.1
+        OR description_similarity > 0.1
+    """)
+    result = ts_conn.execute(query, params)
+    row = result.fetchone()
+    if row:
+        total_count = row[0]
```

Comment:
I deployed the PR to test again and ran the required SQL update script so that I could test at scale.
The query is quite slow with the few playlists we have, so I think we should try to  improve the situation a little bit.

In particular, I'm not convinced we need to do a second DB query to get the total count.
I do understand it is currently used for pagination, but there are a couple ways we can achieve a similar UX without doing two queries each time.

1. don't fetch a total count, and instead expect that you will eventually have a request return a set of results with less than the 25 results requested. That would be a cue to disable the "next" button. In practice that means we are settings totalPageCount in the state to currentpage+1, and if we get not enough results we set it to = currentPage which will disable the "next" button.
The drawback is that the user has to click on next only to find out it is now disabled and we've arrived at the end,
2. Similar to 1, we can request 26 results instead of 25 for each query, and if we get < 26 results we know we have run out of pages and we are on the last one, and we can disable the "next" button. Fetching one extra results removes the drawback of solution 1.
3. Add a new search param to the API endpoints to specifically request a total count. If we only pass this parameter when the query term changes, we avoid getting the count for subsequent pagination requests.

Out of all these, I think # 2 is probably the most straightforward but with a good UX, and should reduce the time these queries take with no visible  drawback for the user.

let me know your thoughts.

Reply:
I have tried optimising the queries by the three methods, but the there isn't much difference in the time taken by the endpoint in this [branch](https://github.com/metabrainz/listenbrainz-server/tree/search-page-improvement). As discussed on IRC, let's keep this part for v2